README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# RGN

<!-- badges: start -->
<!-- badges: end -->

This repository contains the R package for the Robust Gauss-Newton (RGN)
algorithm, which is designed for solving optimization problems with a
sum of least squares objective function.

This R implementation is developed by David McInerney and Michael
Leonard, and is based on original RGN Fortran code developed by Youwei
Qin, Dmitri Kavetski and George Kuczera
(<https://github.com/eachonly/Robust-Gauss-Newton-Algorithm/>).

When using RGN please cite the following articles:

Qin Y, Kavetski D, Kuczera G (2018) A robust Gauss-Newton algorithm for
the optimization of hydrological models: From standard Gauss-Newton to
robust Gauss-Newton. Water Resources Research, 54.
<https://doi.org/10.1029/2017WR022488>

Qin Y, Kavetski D, Kuczera G (2018) A robust Gauss-Newton algorithm for
the optimization of hydrological models: Benchmarking against
industry-standard algorithms. Water Resources Research, 54.
<https://doi.org/10.1029/2017WR022489>

## Installation

You can install the development version of RGN from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ClimateAnalytics/RGN")
```

## Example

The first example is optimisation of a 2D Rosenbrock function:

``` r
library(RGN)
# Example 1: Rosenbrock
simFunc_rosenbrock=function(x) c(1.0-x[1],10.0*(x[2]-x[1]**2))
rgnOut = rgn(simFunc=simFunc_rosenbrock,
             par=c(-1.0,  0.0), lower=c(-1.5, -1.0), upper=c( 1.5,  3.0),
             simTarget=c(0,0))
rgnOut$par #optimal parameters
#> [1] 1 1
rgnOut$value #optimal objective function value
#> [1] 0
```

The second example is calibration of the 5 parameter hydrological model
HYMOD:

``` r
library(RGN)
# Example 2: Hymod
data("BassRiver") # load Bass River hydrological data
rgnOut = rgn(simFunc=simFunc_hymod,
             par=c(400.,0.5,0.1,0.2,0.1),
             lower=c(1.,0.1,0.05,0.000001,0.000001),
             upper=c(1000.,2.,0.95,0.99999,0.99999),
             simTarget=BassRiverData$Runoff.mm.day[365:length(BassRiverData$Date)],
             stateVal=c(100.0,30.0,27.0,25.0,30.0,0.0,0.0,0.0), # initial states for hymod
             nWarmUp=365,                                       # warmup period
             rain=BassRiverData$Rain.mm,                        # precip input
             pet=BassRiverData$ET.mm)                           # PET input
rgnOut$par #optimal parameters
#> [1] 146.7563960   0.3635988   0.1895957   0.9999900   0.7430698
rgnOut$value #optimal objective function value
#> [1] 6840.165
```


NEWS.md
# RGN 1.0.0

* First complete implementation ready for CRAN submission



src/test.f
! this file included as part of workaround to ensure fortran files in src directory are compiled

      subroutine test(a)

      implicit none

      integer,intent(out)::a

      a = 10

      return

      end subroutine test


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(RGN)

test_check("RGN")


tests/testthat/test-hymod.R
test_that('hymod', {

  data("BassRiver")

  tmp = rgn(simFunc=simFunc_hymod,
            par=c(400.,0.5,0.1,0.2,0.1),
            lower=c(1.,0.1,0.05,0.000001,0.000001),
            upper=c(1000.,2.,0.95,0.99999,0.99999),
            simTarget=BassRiverData$Runoff.mm.day[365:length(BassRiverData$Date)],
            nWarmUp=365,
            rain=BassRiverData$Rain.mm,
            pet=BassRiverData$ET.mm,
            control=list(dump=0))
  error=tmp$error;message=tmp$message;par=tmp$par;value=tmp$value;counts=tmp$counts;info=tmp$info

  # expected output based on F90 RGN hymod example
  expect_equal(signif(par,digits=7),
               c(146.7564,0.3635988,0.1895957,0.99999,0.7430698))
  expect_equal(signif(value,digits=7),6840.165)
  expect_equal(counts,367)
  expect_equal(info$nIter,29)
  expect_equal(info$termFlag,2)

})


src/qr.f90
SUBROUTINE Qr_f90(nRow,nCol,a,q,r)
    ! Modified Gram-Schmidt process

   use constantsMod, only: rk

    IMPLICIT NONE

    INTERFACE

    FUNCTION Norm(x)
    USE constantsMod, ONLY: rk
    IMPLICIT NONE
    REAL(rk), DIMENSION(:), INTENT(in) :: x
    REAL(rk) :: Norm
    END FUNCTION Norm

    END INTERFACE

    integer, intent(in) :: nRow, nCol
    REAL(rk), DIMENSION(nRow,nCol), INTENT(in) :: a
    REAL(rk), DIMENSION(nRow,nCol), INTENT(out) :: q
    REAL(rk), DIMENSION(nCol,nCol), INTENT(out) :: r
    REAL(rk), DIMENSION(nRow,nCol) :: a0
    integer :: k,n
    n = nCol
    q(:,:) = 0.0_rk
    r(:,:) = 0.0_rk
    a0(:,:) = a(:,:)

    DO k = 1, n
       r(k,k) = Norm(a0(:,k))
!       q(:,k) = a0(:,k) / r(k,k)
       where(a0(:,k)/=0._rk)  !FIXME_DK: rough fix will need checks+refinement
       q(:,k) = a0(:,k) / r(k,k)
       elsewhere
        q(:,k) = 0._rk
       endwhere
       r(k,k+1:n) = MATMUL(q(:,k), a0(:,k+1:n))
       a0(:,k+1:n) = a0(:,k+1:n) - MATMUL(q(:,k:k), r(k:k,k+1:n))
    END DO

END SUBROUTINE Qr_f90

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

 PURE FUNCTION Norm(x) RESULT(valu)
    ! L2-norm
    USE constantsMod, ONLY: rk
    IMPLICIT NONE
    REAL(rk), DIMENSION(:), INTENT(in) :: x
    REAL(rk) :: valu
    valu = SQRT(SUM(x**2))
  END FUNCTION Norm




src/hymod.f90
!******************************************************************
!
! Purpose: Evolves a single step of the full HYMOD model.
! ---
! Programmer: Youwei Qin and Dmitri Kavetski
! Created: Nov 2013 AD, Uni of Adelaide, Australia.
! Last modified: July 2018 AD, Hohai University, China
! Copyright, Youwei Qin, Dmitri Kavetski, 2018-2023. All rights reserved.
! ---
! Performance
!
! IN:   rain, pet, par, f
!
! OUT:  f
! ---
! Notes
!   Any issues or bugs, please contact the first author(Email:youwei.qin@uon.edu.au)

   SUBROUTINE hymod_f90 (precip, pet, S, &
                     Smax, b, alpha, Ks, Kq, Qs, Qq, Q, err)

   use constantsMod,only:mrk, mik, isoil,iqk1, islo
   IMPLICIT NONE

   INTERFACE

   SUBROUTINE nashCascade(S,k,Vin,Vout)
   use constantsMod,only:mrk, mik
   REAL(mrk),INTENT(inout)::S(:)
   REAL(mrk),INTENT(in)::k
   REAL(mrk),INTENT(in)::Vin
   REAL(mrk),INTENT(out)::Vout
   END SUBROUTINE nashCascade

   SUBROUTINE checkFeasHYMOD(S,Smax,err)
   use constantsMod,only:mrk, mik
   REAL(mrk),INTENT(in)::S(:)
   REAL(mrk),INTENT(in)::Smax
   INTEGER(mik),INTENT(out)::  err
   END SUBROUTINE checkFeasHYMOD

    END INTERFACE

    !dummy parameters
   REAL(mrk),INTENT(in):: precip, pet           ! forcings
   INTEGER(mik),INTENT(out)::  err
   REAL(mrk),INTENT(inout)::S(5)
   REAL(mrk),INTENT(in)::Smax,b,alpha,Ks,Kq     ! parameters
   REAL(mrk),INTENT(out)::Qs,Qq,Q               ! responses

   !local parameters
   CHARACTER(*),PARAMETER::procnam="evolveHYMOD"
   REAL(mrk)::Uo,Uq,Us,Ue

  err = 0

   ! Start procedure here
    CALL checkFeasHYMOD(S,Smax,err)
   !The first part is to calculate the sum of direct flow and saturated flow,U0
   CALL PDM(S(isoil),Smax,b,precip,Uo,err)

   Ue=pet                                       !     - ET at max (eg, Wagener HESS)
   Ue=min(Ue,S(isoil))
   S(isoil)=S(isoil)-Ue                         !       final soil storage
   Uq=alpha*Uo; Us=Uo-Uq                        ! *** quick vs slow flow routing
   CALL nashCascade(S(iqk1:),Kq,Uq,Qq)      ! quick flow (Cascade Of The Nash)
   CALL nashCascade(S(islo:islo),Ks,Us,Qs)  ! slow flow (single linear reservoir)
   Q=Qs+Qq

      END SUBROUTINE hymod_f90



   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

   PURE SUBROUTINE PDM(S,Smax,b,P,Q,err)
   ! Purpose: Single time step of the Probability Distribution Model (PDM).
   !          (finite non-leaky tank - soft threshold)
   ! ---
   ! Programmer: Dmitri Kavetski, Augusta-6 Glenelg
   ! ---
   ! Motivated by Matlab code 'ModPdm01.m' of Hoshin V. Gupta (9/11/2005)
   ! ---
   ! INPUTS
   !   S       = initial storage
   !   Smax    = maximum storage parameter
   !   b       = exponent parameter
   !   P       = precipitation for the time-step
   ! OUTPUTS
   !   S       = final storage
   !   Q       = model outflow (spillage + soft)
   ! ---
   ! Notes
   ! 1. Smax=Cmax/(1+Bexp)=Cmax*Bconst and Bconst=1/(1+Bexp)
   ! 2. Uses S as the primary state variable (faster and cleaner)
   ! 3. HVG used b=log(1-(Bpar/2))/log(0.5) to convert scaled B [0,2] to unscaled b [0,inf)
   ! ---
   USE constantsMod,ONLY:zero,half,one, mrk, mik
   IMPLICIT NONE
   ! dummies
   REAL(mrk),INTENT(inout)::S
   REAL(mrk),INTENT(in)::Smax,b,P
   REAL(mrk),INTENT(out)::Q
   INTEGER(mik),INTENT(inout)::err
   ! locals
   REAL(mrk),PARAMETER::Smin=zero,tolMBE=1.e-14_mrk
   REAL(mrk)::Sbeg,frac,OV1,OV2,Pinf,eMB,net
   REAL(mrk)::b1,ob1,Hmax,Hbeg,Hend
   CHARACTER(*),PARAMETER::procnam="PDM"
   ! Start procedure here
   IF(S>Smax)THEN
     err=-10; RETURN
   ELSEIF(S<Smin)THEN
     err=-10; RETURN
   ELSEIF(b<zero)THEN
     err=-10; RETURN
   ENDIF
   Sbeg=S; OV2=zero
  ! Moore's description, D'z implementation based on Gupta's Matlab
   b1=b+one; ob1=one/b1; Hmax=Smax*b1  ! max height
   frac=one-(one-S/Smax)**ob1          ! active fraction
   Hbeg=Hmax*frac                      ! initial height
   OV2=MAX(Hbeg+P-Hmax,zero)           ! spillage
   Pinf=P-OV2                          ! P that does not spill
   Hend=MIN(Hbeg+Pinf,Hmax)            ! final height
   S=Smax*(one-(one-Hend/Hmax)**b1)    ! final storage
   OV1=MAX(Sbeg+Pinf-S,zero)           ! soft outflow
   Q=OV1+OV2                           ! total outflow
   IF(S>Smax)THEN
     err=-10; RETURN
   ELSEIF(S<Smin)THEN
     err=-10; RETURN
   ENDIF
   net=P-Q
   eMB=Sbeg+net-S                      ! S-balance discrEpency
   IF(ABS(eMB)>tolMBE*MAX(Smax,P,Q))THEN
     err=-10; RETURN
   ENDIF
   ! End procedure here
   ENDSUBROUTINE PDM

   !----------------------------------------------------
   PURE SUBROUTINE nashCascade(S,k,Vin,Vout)
   ! Purpose: Executes a single time step of the Nash Cascade Model.
   !          (Series of infinite linear leaky tanks)
   ! ---
   ! Programmer: Dmitri Kavetski, Augusta-6 Glenelg
   ! ---
   ! Motivated by Matlab code 'ModNash.m' of Hoshin V. Gupta (9/18/2005)
   ! ---
   ! INPUTS
   !   S     = Initial storage in individual tanks
   !   k     = Leakage rate of the tanks
   !   Vin   = Input to first tank in the series
   ! OUTPUTS
   !   Vout  = Outflow from last tank in series
   !   S     = Final storage in individual tanks
   ! ---
   USE constantsMod,ONLY:zero,half,one, mrk, mik
   IMPLICIT NONE
   ! dummies
   REAL(mrk),INTENT(inout)::S(:)
   REAL(mrk),INTENT(in)::k
   REAL(mrk),INTENT(in)::Vin
   REAL(mrk),INTENT(out)::Vout
   ! locals
   INTEGER(mik)::i,n
   REAL(mrk)::q(0:SIZE(S))
   ! Start procedure here
   n=SIZE(S); q(0)=Vin
   DO i=1,n
     q(i)=k*S(i); S(i)=S(i)+q(i-1)-q(i)
   ENDDO
   Vout=q(n)
   ! End procedure here
   ENDSUBROUTINE nashCascade

   !----------------------------------------------------
    PURE SUBROUTINE checkFeasHYMOD(S,Smax,err)
    ! Purpose: Checks feasibility of HYMOD states and parameters.
    USE constantsMod,ONLY:zero, mrk, mik, isoil
    IMPLICIT NONE
    ! dummies
    REAL(mrk),INTENT(in)::S(:),Smax
    INTEGER(mik),INTENT(out)::err
    ! locals
    CHARACTER(*),PARAMETER::procnam="checkFeasHYMOD"
    CHARACTER(*),PARAMETER::fmt1='(a,es15.8,a,es15.8,a)'
    CHARACTER(*),PARAMETER::fmt2='(a,i0,a,es15.8,a)'
    ! Start procedure here
    IF(S(isoil)>Smax)THEN
      err=-10; RETURN
    ELSEIF(ANY(S<zero))THEN
      err=-10; RETURN
    ENDIF
    ! End procedure here
    ENDSUBROUTINE checkFeasHYMOD
   !----------------------------------------------------


R/constantsMod.R
#******************************************************************
# Purpose: modules for some most widely used constants

# Author: George Kuczera
# Copyright, George Kuczera, 2005-2010. All rights reserved.
# Modified by Michael Leonard Dec 2022 - Without permission

ik4=4  # SELECTED_INT_KIND(9)   # Integer kind
ik=8   # SELECTED_INT_KIND(9)   # Integer kind
rk4=4  # Real*4 kind
rk=8   # Real kind

# Machine constants
hugeRe  = .Machine$double.xmax    # largest real on machine
tinyRe  = .Machine$double.xmin    # smallest real on machine
epsRe   = .Machine$double.eps     # smallest real epsilon on machine
minExp  = .Machine$double.min.exp # smallest exponent
maxExp  = .Machine$double.max.exp # biggest exponent
hugeInt = .Machine$integer.max    # largest integer on machine

# Other constants
zero=0.0; half=0.5; one=1.0; two=2.0; four=4.0
pi = 3.141592653589793238462643383279502884197
twoPi = 6.283185307179586476925286766559005768394

  # Script constants and structures
maxScriptLines=700; nCharScriptLine=60; scriptSize=maxScriptLines*nCharScriptLine; codeSize=3*scriptSize; maxnScriptProc=40

scriptProcType=list()
scriptProcType$nCode=0
scriptProcType$name=""
scriptProcType$nameUC=""
scriptProcType$description=""
scriptProcType$formula=""
scriptProcType$code=rep(0,codeSize)



src/constantsMod.f90
!******************************************************************
!
! Purpose: modules for some most widely used constants

! Author: George Kuczera
! Email: george.kuczera@newcastle.edu.au
! School of Engineering
! University of Newcastle
! Callaghan, NSW, 2308, Australia
!
! Copyright, George Kuczera, 2005-2010. All rights reserved.
!
MODULE constantsMod
   IMPLICIT NONE
   SAVE
   PUBLIC

   INTEGER, PARAMETER :: mik=8      ! SELECTED_INT_KIND(9)   ! Integer kind
   INTEGER, PARAMETER :: mrk=8      ! Real kind

   INTEGER(mik),PARAMETER::isoil=1,islo=2,iqk1=3,iqk2=4,iqk3=5

   INTEGER, PARAMETER :: ik4=4   ! SELECTED_INT_KIND(9)   ! Integer kind
   INTEGER, PARAMETER :: ik=8   ! SELECTED_INT_KIND(9)   ! Integer kind
   INTEGER, PARAMETER :: rk4=4   ! Real*4 kind
   INTEGER, PARAMETER :: rk=8   ! Real kind
   !
   ! Machine constants
   REAL(rk), PARAMETER :: hugeRe = HUGE(1.0_rk)           ! largest real on machine
   REAL(rk), PARAMETER :: tinyRe = TINY(1.0_rk)           ! smallest real on machine
   REAL(rk), PARAMETER :: epsRe = EPSILON(1.0_rk)         ! smallest real epsilon on machine
   INTEGER(ik), PARAMETER :: minExp = MINEXPONENT(1.0_rk) ! smallest exponent
   INTEGER(ik), PARAMETER :: maxExp = MAXEXPONENT(1.0_rk) ! biggest exponent
   INTEGER(ik), PARAMETER :: hugeInt = HUGE(1_ik)            ! largest integer on machine
   !
   ! Other constants
   REAL(rk), PARAMETER :: zero=0.0_rk, half=0.5_rk, one=1.0_rk, two=2.0_rk, four=4.0_rk
   REAL(rk), PARAMETER :: pi = 3.141592653589793238462643383279502884197_rk
   REAL(rk), PARAMETER :: twoPi = 6.283185307179586476925286766559005768394_rk
   !
   ! Script constants and structures
   INTEGER(ik), PARAMETER :: maxScriptLines=700, nCharScriptLine=60, &
   scriptSize=maxScriptLines*nCharScriptLine, codeSize=3*scriptSize, maxnScriptProc=40
   TYPE scriptProcType
      INTEGER(ik) :: nCode
      CHARACTER(20) :: name, nameUC
      CHARACTER(50) :: description
      CHARACTER(scriptSize) :: formula
      INTEGER(ik4) :: code(0:codeSize)
   END TYPE scriptProcType

END MODULE constantsMod



R/simHymod.R
#' @title hymod simulation
#' @description Simulation of hymod rainfall-runoff model
#' @param x parameter values
#' @param nWarmUp length of warmup period
#' @param rain precipitation input (mm/day)
#' @param pet potential evapotranspiration (mm/day)
#' @param stateVal (optional) initial states
#' @return Vector of simulated runoff
#' @export
simFunc_hymod = function(x,rain,pet,nWarmUp,
                         stateVal=c(100.0,30.0,27.0,25.0,30.0,0.0,0.0,0.0)){
  procnam = 'simFunc_hymod'
  S = vector(length=length(x))
  flexS=TRUE         # Allow fix of Smax
  #Assign parameters
  Smax=x[1]; b=x[2]; alpha=x[3]; Ks=x[4]; Kq=x[5]
  # Initialize surfacewater storages and baseflowstorage, and the initial
  # storage of C1,C2,C3
  S[1]=stateVal[1];S[2]=stateVal[2];S[3]=stateVal[3];S[4]=stateVal[4]
  S[5]=stateVal[5]; Qs=stateVal[6];Qq=stateVal[7];Q=stateVal[8]
  # check feas of state wrt pars unless flexi-state
  if ((S[1]>Smax)&(!flexS)){
    message=paste0("f-",procnam,"/Soil moisture exceeds")
    error=-10
    return(list(error=error,message=message))
  }
  if(flexS){
    if(S[1]>Smax){S[1]=Smax}
  }
  nSim = length(rain)
  Qvec = vector(length = nSim)
  for (i in 1:nSim){
    qsimf<-.Fortran("hymod_f90",
                    precip=as.double(rain[i]), pet=as.double(pet[i]),
                    S=as.double(S), Smax=as.double(Smax), b=as.double(b),
                    alpha=as.double(alpha), Ks=as.double(Ks),
                    Kq=as.double(Kq), Qs=as.double(1), Qq=as.double(1),
                    Q=as.double(1), err = as.integer(1))
    Qvec[i] = qsimf$Q
    S = qsimf$S
  }
  return(Qvec[nWarmUp:nSim])
}
